src/main/java/org/veupathdb/service/eda/ds/core/AbstractPlugin.java

```diff
@@ -611,4 +650,147 @@ public String getRBinListAsString(List<String> labels) {
+        studyVocabListRString = studyVocabListRString + studyVocabRString;
+        first = false;
+      } else {
+        studyVocabListRString = studyVocabListRString + "," + studyVocabRString;
+      }
+    }
+    
+    return studyVocabListRString + "))";
+  }
+
+  public boolean hasStudyDependentVocabulary(DynamicDataSpecImpl dataSpec) {
+    PluginUtil util = getUtil();
+
+    if (dataSpec.isCollectionSpec()) {
+      return util.getHasStudyDependentVocabulary(dataSpec.getCollectionSpec());
+    } else if (dataSpec.isVariableSpec()) {
+      return util.getHasStudyDependentVocabulary(dataSpec.getVariableSpec());
+    } else {
+      return false;
+    }
+  }
+
+  public List<DynamicDataSpecImpl> findDataSpecsWithStudyDependentVocabs(Map<String, DynamicDataSpecImpl> dataSpecs) {
+    List<DynamicDataSpecImpl> dataSpecsWithStudyDependentVocabs = new ArrayList<>();
+
+    for (DynamicDataSpecImpl dataSpec : dataSpecs.values()) {
+      if (hasStudyDependentVocabulary(dataSpec)) {
+        dataSpecsWithStudyDependentVocabs.add(dataSpec);
+      } 
+    }
+            
+    return dataSpecsWithStudyDependentVocabs;
+  }
+
+  public List<DynamicDataSpecImpl> findVariableSpecsWithStudyDependentVocabs(Map<String, VariableSpec> varSpecs) {
+    Map<String, DynamicDataSpecImpl> dataSpecs = varMapToDynamicDataMap(varSpecs);
+
+    return findDataSpecsWithStudyDependentVocabs(dataSpecs);
+  }
+
+  public List<DynamicDataSpecImpl> findCollectionSpecsWithStudyDependentVocabs(Map<String, CollectionSpec> collectionSpecs) {
+    Map<String, DynamicDataSpecImpl> dataSpecs = collectionMapToDynamicDataMap(collectionSpecs);
+
+    return findDataSpecsWithStudyDependentVocabs(dataSpecs);
+  }
+
+  public String getDynamicDataSpecEntityId(DynamicDataSpecImpl dataSpec) {
+    if (dataSpec.isCollectionSpec()) {
+      return dataSpec.getCollectionSpec().getEntityId();
+    } else if (dataSpec.isVariableSpec()) {
+      return dataSpec.getVariableSpec().getEntityId();
+    } else {
+      return null;
+    }
+  }
+
+  public String getDynamicDataSpecId(DynamicDataSpecImpl dataSpec) {
+    if (dataSpec.isCollectionSpec()) {
+      return dataSpec.getCollectionSpec().getCollectionId();
+    } else if (dataSpec.isVariableSpec()) {
+      return dataSpec.getVariableSpec().getVariableId();
+    } else {
+      return null;
+    }
+  }
+
+  public String getRMegastudyAsString(String compressedDataHandle, Map<String, DynamicDataSpecImpl> dataSpecs) {
+    PluginUtil util = getUtil();
+
+    // find and validate variables w study specific vocabs
+    List<DynamicDataSpecImpl> dataSpecsWithStudyDependentVocabs = findDataSpecsWithStudyDependentVocabs(dataSpecs);
+    List<String> entities = dataSpecsWithStudyDependentVocabs.stream().map(data -> getDynamicDataSpecEntityId(data)).toList();
+    boolean allEqualEntities = entities.isEmpty() || Collections.frequency(entities, entities.get(0)) == entities.size();
+    if (!allEqualEntities) { 
+      // TODO get mad 
+    }
```

Comment:
Would it ever be valid to pass duplicate data specs in here? If we don't want to do this validation here we could state that it's a precondition that data specs are not duplicate and let the validation live upstream.

Reply:
hmm. probably we wont ever see duplicate data specs, but im not sure we need to explicitly prevent against it? the validation i was thinking of was more that we actually had to do anything. the circumstances under which we need to impute anything is kind of specific, so it makes sense of those conditions arent met that we simply dont do anything. i just need to remember/ collect a detailed list of what those conditions are..